src/Core/Domain/CartRule/ValueObject/CartRuleAction.php

```diff
@@ -24,61 +24,60 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  */
 
-namespace PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\CartRuleAction;
+declare(strict_types=1);
 
-use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\GiftProduct;
-use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\PercentageDiscount;
-use PrestaShop\PrestaShop\Core\Domain\ValueObject\Money;
+namespace PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject;
 
-/**
- * Cart rule action that gives free shipping.
- * It cannot have percentage or amount discount.
- * It can optionally have gift product.
- */
-final class FreeShippingAction implements CartRuleActionInterface
+use PrestaShop\PrestaShop\Core\Domain\CartRule\Exception\CartRuleConstraintException;
+
+class CartRuleAction
 {
     /**
-     * @var GiftProduct|null
+     * @var bool
      */
-    private $giftProduct;
+    private $freeShipping;
 
     /**
-     * @param GiftProduct|null $giftProduct
+     * @var GiftProduct|null
      */
-    public function __construct(?GiftProduct $giftProduct = null)
-    {
-        $this->giftProduct = $giftProduct;
-    }
+    private $giftProduct;
 
     /**
-     * {@inheritdoc}
+     * @var Discount|null
      */
-    public function isFreeShipping(): bool
-    {
-        return true;
-    }
+    private $discount;
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getPercentageDiscount(): ?PercentageDiscount
-    {
-        return null;
+    public function __construct(
+        bool $freeShipping,
+        ?GiftProduct $giftProduct = null,
+        ?Discount $discount = null
+    ) {
+        if ($freeShipping || $giftProduct || $discount) {
```

Comment:
I'd inverse the condition here to avoid having a `return` inside a `__construct` , but it is not a big deal

Reply:
Haha great minds 😅 I had the same comment!

@zuk3975 the PR is validated by both of us, if you wanna handle this feedback go for it, if not you can merge the PR (partial PR with green tests!!)